cuqi/mcmc/_sampler.py

```diff
@@ -0,0 +1,324 @@
+from abc import ABC, abstractmethod
+import os
+import sys
+import numpy as np
+import pickle as pkl
+import cuqi
+from cuqi.samples import Samples
+from progressbar import progressbar
+
+class SamplerNew(ABC):
+    """ Abstract base class for all samplers.
+    
+    Provides a common interface for all samplers. The interface includes methods for sampling, warmup and getting the samples in an object oriented way.
+
+    Samples are stored in a list to allow for dynamic growth of the sample set. Returning samples is done by creating a new Samples object from the list of samples.
+
+    """
+
+    def __init__(self, target: cuqi.density.Density, initial_point=None, callback=None):
```

Comment:
We discussed allowing an empty target at some point, but I forgot for what reason. Just a reminder here:
- `self.dim` is inferred from target
- Constructors of derived classes also need target to initialize some variables, for example in  `MALANew`, we need `self.current_target_eval = self.target.logd(self.current_point)`

Reply:
Added to #378